### PR TITLE
gateway: Add Access Log for Object Copy

### PR DESCRIPTION
**Pull request by windkit**

### CODE REVIEW

**Comment by yosukehara** [comment]
LGTM to me. @mocchira Please review it.

File: apps/leo_gateway/include/leo_gateway.hrl [hunk 1]
```diff
@@ -445,6 +445,25 @@
               })
             %% ?notify_metrics(<<"PUT">>,_Bucket,_Size)
         end).
+-define(access_log_copy(_Bucket,_Path,_Size,_Response,_Begin),
+        begin
+            _Latency = erlang:round((leo_date:clock() - _Begin) / 1000),
```
Comment (mocchira) [inline apps/leo_gateway/include/leo_gateway.hrl]: Assign a variable to leo_date:clock and reference it here and the following code where leo_date:clock is called in order to reduce os:timestamp calls.

File: apps/leo_gateway/include/leo_gateway.hrl [hunk 2]
```diff
@@ -576,6 +595,42 @@
                                       ]}
               })
         end).
+-define(access_log_bucket_getacl(_Bucket, _Response, _Begin),
+        begin
+            _Latency = erlang:round((leo_date:clock() - _Begin) / 1000),
```
Comment (mocchira) [inline apps/leo_gateway/include/leo_gateway.hrl]: Same with the above comment.

File: apps/leo_gateway/include/leo_gateway.hrl [hunk 3]
```diff
@@ -576,6 +595,42 @@
...
+                                       leo_date:clock(),
+                                       _Response,
+                                       _Latency,
+                                       ""
+                                      ]}
+              })
+        end).
+-define(access_log_bucket_getacl(_Bucket, _CannedACL, _Response, _Begin),
+        begin
+            _Latency = erlang:round((leo_date:clock() - _Begin) / 1000),
```
Comment (mocchira) [inline apps/leo_gateway/include/leo_gateway.hrl]: Same as above

File: apps/leo_gateway/src/leo_gateway_s3_api.erl [hunk 4]
```diff
@@ -603,44 +632,53 @@ put_object(Directive, Req, Key, #req_params{handler = ?PROTO_HANDLER_S3,
...
                                                      meta = CMetaBin,
                                                      dsize = Size,
                                                      msize = byte_size(CMetaBin),
                                                      bucket_info = BucketInfo}) of
         {ok, _ETag} when Directive == ?HTTP_HEAD_X_AMZ_META_DIRECTIVE_COPY ->
-            ?access_log_put(BucketName, Key, Size, ?HTTP_ST_OK, BeginTime),
+            ?access_log_copy(BucketName, SrcDst, Size, ?HTTP_ST_OK, BeginTime),
             resp_copy_obj_xml(Req, Meta);
         {ok, _ETag} when Directive == ?HTTP_HEAD_X_AMZ_META_DIRECTIVE_REPLACE ->
+            ?access_log_copy(BucketName, SrcDst, Size, ?HTTP_ST_OK, BeginTime),
```
Comment (windkit) [inline apps/leo_gateway/src/leo_gateway_s3_api.erl]: @mocchira Actually here I don't understand why there would be a delete after the copy when `x-amz-metadata-directive` is set to `REPLACE`

Comment (mocchira) [inline apps/leo_gateway/src/leo_gateway_s3_api.erl]: Right, it should be remained so please fix the problem on this PR. I will file the issue later.

Comment (mocchira) [inline apps/leo_gateway/src/leo_gateway_s3_api.erl]: Filed here: https://github.com/leo-project/leofs/issues/784

**Comment by windkit** [comment]
@mocchira I have updated the PR, please review

**Comment by mocchira** [comment]
LGTM.

**Pull request merged**